compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs

```diff
@@ -460,17 +459,17 @@ generatePackageMap version mbProjRoot userPkgDbs = do
             let dalfs = filter ((== ".dalf") . takeExtension) allFiles
             forM dalfs $ \dalf -> do
                 dalfPkgOrErr <- readDalfPackage dalf
-                let baseName = takeBaseName dalf
-                let getUnitId pkg
-                -- If we use data-dependencies we can end up with multiple DALFs for daml-prim/daml-stdlib
-                -- one per version. The one shipped with the SDK is called daml-prim.dalf and daml-stdlib-$VERSION.dalf
-                -- and have the same unit ids, so we do not need to strip package ids.
-                -- The one coming from daml-prim will be called daml-prim-$PKGID.dalf daml-stdlib-$PKGID.dalf
-                -- To avoid collisions, we include this hash in the unit id so we also don’t want to strip
-                -- package ids here.
-                        | "daml-prim" `isPrefixOf` baseName = (stringToUnitId baseName, pkg)
-                        | "daml-stdlib" `isPrefixOf` baseName = (stringToUnitId baseName, pkg)
-                        | otherwise = ( stringToUnitId $ parseUnitId baseName $ LF.dalfPackageId pkg, pkg)
+                let getUnitId pkg = case LF.packageMetadataFromFile dalf (LF.extPackagePkg $ LF.dalfPackagePkg pkg) (LF.dalfPackageId pkg) of
```

Comment:
This is a bit tough to parse, but it might also just be because of GitHub's (lack of) syntax highlighting in diffs. Or it's due to the `case` being so far from the patterns/guards. Perhaps it could some refactoring to make it clearer.

Reply:
I completely agree! Didn’t seem quite so bad when writing it but looking at it now, this looks completely unreadable. I’ve moved `getUnitId` to a `where`-clause and moved the comment above the logic which is hopefully easier to parse.